frontend/src/metabase/visualizations/shared/utils/data.ts

```diff
@@ -0,0 +1,248 @@
+  return Array.from(keys).reduce<MetricDatum>((datum, metricKey) => {
+    const leftValue = left[metricKey];
+    const rightValue = right[metricKey];
+
+    if (typeof leftValue === "number" || typeof rightValue === "number") {
+      datum[metricKey] = (leftValue ?? 0) + (rightValue ?? 0);
+    } else {
+      datum[metricKey] = null;
+    }
+
+    return datum;
+  }, {});
+};
+
+const groupDataByDimensions = (
+  rows: RowValues[],
+  chartColumns: ChartColumns,
+  allMetrics: ColumnDescriptor[],
+  columnFormatter: ColumnFormatter,
+): GroupedDataset => {
+  const { dimension } = chartColumns;
+
+  const groupedData = new Map<RowValue, GroupedDatum>();
+
+  for (const row of rows) {
+    const dimensionValue = row[dimension.index];
+
+    const datum = groupedData.get(dimensionValue) ?? {
+      dimensionValue,
+      metrics: {},
+    };
+
+    const rowMetrics = allMetrics.reduce<MetricDatum>((datum, metric) => {
+      datum[metric.column.name] = getMetricValue(row[metric.index]);
+      return datum;
+    }, {});
+
+    datum.metrics = sumMetrics(rowMetrics, datum.metrics);
+
+    if ("breakout" in chartColumns) {
+      const breakoutName = columnFormatter(
+        row[chartColumns.breakout.index],
+        chartColumns.breakout.column,
+      );
+
+      datum.breakout = {
+        ...datum.breakout,
+        [breakoutName]: sumMetrics(
+          rowMetrics,
+          datum.breakout?.[breakoutName] ?? {},
+        ),
+      };
+    }
+
+    groupedData.set(dimensionValue, datum);
+  }
+
+  return Array.from(groupedData.values());
+};
+
+export const getGroupedDataset = (
+  data: TwoDimensionalChartData,
+  chartColumns: ChartColumns,
+  columnFormatter: ColumnFormatter,
+): GroupedDataset => {
+  // We are grouping all metrics because they are used in chart tooltips
+  const allMetricColumns = data.cols
+    .filter(isMetric)
+    .map(column => column.name);
+
+  const allMetricDescriptors = getColumnDescriptors(
+    allMetricColumns,
+    data.cols,
+  );
+
+  return groupDataByDimensions(
+    data.rows,
+    chartColumns,
+    allMetricDescriptors,
+    columnFormatter,
+  );
+};
+
+export const trimData = (
+  dataset: GroupedDataset,
+  valuesLimit: number,
+): GroupedDataset => {
+  if (dataset.length <= valuesLimit) {
+    return dataset;
+  }
+
+  const groupStartingFromIndex = valuesLimit - 1;
+  const result = dataset.slice();
+  const dataToGroup = result.splice(groupStartingFromIndex);
+
+  const groupedDatumDimensionValue =
+    dataToGroup.length === dataset.length
+      ? t`All values (${dataToGroup.length})`
+      : t`Other (${dataToGroup.length})`;
+
+  const groupedValuesDatum = dataToGroup.reduce(
+    (groupedValue, currentValue) => {
+      groupedValue.metrics = sumMetrics(
+        groupedValue.metrics,
+        currentValue.metrics,
+      );
+
+      Object.keys(currentValue.breakout ?? {}).map(breakoutName => {
+        groupedValue.breakout ??= {};
```

Comment:
Actually, it is completely valid to have data without breakouts so probably I should not have `breakout: {}` in the initial value since it will be just an empty object for charts without a breakout. What do you think?

Reply:
That sounds better since `breakout` is optional in `GroupDatum` so using `Required<GroupDatum>` is kind of weird in a way that sometimes we expect `breakout` to be there and sometimes we don't.